### PR TITLE
Add accelerator table

### PR DESCRIPTION
**Pull request by dpanshug**
Closes: #1373

## Description
Added Accelerator profile table
- [X] Render the following columns: Device identifier, display name/description, ennoblement switch, date last modified, action button
- [X] Columns (except for action column) must be sortable
- [X] include help icons where present in the docs
- [X] should include pagination
- [X] ellipsis action button should include: edit, delete
- [X] hook to fetch accelerator CRDs

image

## How Has This Been Tested?
1. Add accelerator profile CR's to your cluster.
2. Login to ODH dashboard with clusteradmin
3. Navigate to settings < Accelerator profiles
4. You can see the Accelerator table, with all it's fields, sortable columns, and action buttons.

## Test Impact
Added storybook component and integration test.

## Request review criteria:

Self checklist (all need to be checked):
- [x] The developer has manually tested the changes and verified that the changes work
- [x] Commits have been squashed into descriptive, self-contained units of work (e.g. 'WIP' and 'Implements feedback' style messages have been removed)
- [x] Testing instructions have been added in the PR body (for PRs involving changes that are not immediately obvious).
- [x] The developer has added tests or explained why testing cannot be added (unit tests & storybook for related changes)

If you have UI changes: 

- [x] Included any necessary screenshots or gifs if it was a UI change.
- [x] Included tags to the UX team if it was a UI/UX change (find relevant UX in the [SMEs](https://github.com/opendatahub-io/odh-dashboard/tree/main/docs/smes.md) section).

After the PR is posted & before it merges:
- [ ] The developer has tested their solution on a cluster by using the image produced by the PR to `main`

@yannnz

### CODE REVIEW

File: frontend/src/pages/acceleratorProfiles/components/TableData.tsx [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+    label: 'Name',
+    sortable: (a, b) => a.spec.displayName.localeCompare(b.spec.displayName),
+    width: 30,
+  },
+  {
+    field: 'identifier',
+    label: 'Identifier',
+    sortable: (a, b) => a.spec.identifier.localeCompare(b.spec.identifier),
+    info: {
+      tooltip: 'The resource identifier of the accelerator device.',
```
Comment (Gkrumbach07) [inline frontend/src/pages/acceleratorProfiles/components/TableData.tsx:16]: use popover

File: frontend/src/pages/acceleratorProfiles/components/TableData.tsx [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+    },
+  },
+  {
+    field: 'enablement',
+    label: 'Enable',
+    sortable: (a, b) => Number(a.spec.enabled ?? false) - Number(b.spec.enabled ?? false),
+    info: {
+      tooltip: 'Indicates whether the accelerator profile is available for new resources.',
```
Comment (Gkrumbach07) [inline frontend/src/pages/acceleratorProfiles/components/TableData.tsx:24]: popover

File: frontend/src/pages/acceleratorProfiles/screens/list/AcceleratorProfilesTableRow.tsx [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+    <Td dataLabel="Enable">
+      <AcceleratorProfileEnableToggle
+        enabled={accelerator.spec.enabled}
+        name={accelerator.spec.displayName}
+      />
+    </Td>
+    <Td dataLabel="Last modified">
+      {accelerator.metadata.annotations?.['opendatahub.io/modified-date'] ? (
+        <Timestamp
+          date={new Date(accelerator.metadata.annotations['opendatahub.io/modified-date'])}
```
Comment (Gkrumbach07) [inline frontend/src/pages/acceleratorProfiles/screens/list/AcceleratorProfilesTableRow.tsx:33]: invalid dates render as _now_. We should render invalid dates as `--`

File: frontend/src/pages/acceleratorProfiles/components/TableData.tsx [hunk 4]
```diff
@@ -0,0 +1,41 @@
+import { SortableData } from '~/components/table';
```
Comment (Gkrumbach07) [inline frontend/src/pages/acceleratorProfiles/components/TableData.tsx:1]: this should not be a `tsx` file. put this in a file named `const.ts`. Also instead of using a components folder at the root.  place all relevant files in their respective screen folders. Unless that component. will be used in both the edit page and list page

```
page/acceleratorProfiles
   - screens
      - list
         - AcceleratorProfiles.tsx
         - AcceleratorProfilesTable.tsx
         - AcceleratorProfilesTableRow.tsx
         - AcceleratorProfileEnableToggle.tsx
         - const.ts
```

File: frontend/src/pages/acceleratorProfiles/components/TableData.tsx [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+import { AcceleratorKind } from '~/k8sTypes';
+
+export const columns: SortableData<AcceleratorKind>[] = [
+  {
+    field: 'name',
+    label: 'Name',
+    sortable: (a, b) => a.spec.displayName.localeCompare(b.spec.displayName),
+    width: 30,
```
Comment (Gkrumbach07) [inline frontend/src/pages/acceleratorProfiles/components/TableData.tsx:9]: why is width needed here?

File: frontend/src/pages/acceleratorProfiles/screens/list/AcceleratorProfilesTable.tsx [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+};
+
+const AcceleratorProfilesTable: React.FC<AcceleratorProfilesTableProps> = ({ accelerators }) => (
+  <Table
+    enablePagination
+    data={accelerators}
+    columns={columns}
+    emptyTableView={'No projects match your filters.'}
+    rowRenderer={(accelerator) => (
+      <AcceleratorProfilesTableRow key={accelerator.metadata.uid} accelerator={accelerator} />
```
Comment (Gkrumbach07) [inline frontend/src/pages/acceleratorProfiles/screens/list/AcceleratorProfilesTable.tsx:18]: our types say uid can be undefined. please use `name`

**Review by Gkrumbach07 — CHANGES_REQUESTED** [review]
This is a empty file
[frontend/src/pages/acceleratorProfiles/components/SearchField.tsx](https://github.com/opendatahub-io/odh-dashboard/pull/2069/files#diff-7be3f07123886d27b34f7b3c18d542d423fe91a2f159e920eccdcc0a0e42fd60)

File: frontend/src/__tests__/integration/pages/acceleratorProfiles/AcceleratorProfiles.stories.tsx [hunk 7]
```diff
@@ -48,3 +60,14 @@ export const EmptyStateNoAcceleratorProfile: StoryObj = {
...
 };
+
+export const ListAcceleratorProfiles: StoryObj = {
+  render: Template,
+
+  play: async ({ canvasElement }) => {
+    // load page and wait until settled
+    const canvas = within(canvasElement);
+    await canvas.findByText('Test Accelerator', undefined, { timeout: 5000 });
+    await canvas.findByText('TensorRT', undefined, { timeout: 5000 });
```
[the remark this answers is not in the document]
Comment (dpanshug) [reply frontend/src/__tests__/integration/pages/acceleratorProfiles/AcceleratorProfiles.stories.tsx:71]: I see, so adding timeout for only first check is necessary. Removing from second check.

File: frontend/src/pages/acceleratorProfiles/components/TableData.tsx (hunk 5 above)
Comment (dpanshug) [reply frontend/src/pages/acceleratorProfiles/components/TableData.tsx:9]: Width is needed so that when the description is large it won't take extra space and can be truncated easily. 

without width
image

Comment (Gkrumbach07) [reply frontend/src/pages/acceleratorProfiles/components/TableData.tsx:9]: The mocks looks like they say to wrap the text
https://www.sketch.com/s/d8550d2f-fae4-44fd-82df-9f400563d459/a/27bop9y

`<Td>` has a modifier prop where you can control this
``` ts
modifier?: "wrap" | "nowrap" | "breakWord" | "fitContent" | "truncate" | undefined
```

Comment (dpanshug) [reply frontend/src/pages/acceleratorProfiles/components/TableData.tsx:9]: But what about when the description is long. Should we let it take the table space? 

<details>
<summary>With Truncate component and width</summary>



</details>

<details>
<summary>With wrap modifier</summary>

Screenshot from 2023-11-08 11-49-59
</details>

<details>
<summary>With truncate modifier</summary>

Screenshot from 2023-11-08 11-52-56
</details>

@yannnz What do you think? What should we go ahead with?

Comment (yannnz) [reply frontend/src/pages/acceleratorProfiles/components/TableData.tsx:9]: Is the first option limit the width of the description and applies truncate? If that is how it visually looks like, I will vote for the first one.

Comment (dpanshug) [reply frontend/src/pages/acceleratorProfiles/components/TableData.tsx:9]: First option limits the width of whole column `Name`

Comment (yannnz) [reply frontend/src/pages/acceleratorProfiles/components/TableData.tsx:9]: > First option limits the width of whole column `Name`

It makes sense to limit the whole column for both long name and long description with Truncate. So long names and description won't take the entire space of the table.

Comment (dpanshug) [reply frontend/src/pages/acceleratorProfiles/components/TableData.tsx:9]: So I believe we can consider this conversation resolved, as no further changes are needed. cc @Gkrumbach07

File: frontend/src/pages/acceleratorProfiles/screens/list/AcceleratorProfilesTableRow.tsx (hunk 3 above)
Comment (dpanshug) [reply frontend/src/pages/acceleratorProfiles/screens/list/AcceleratorProfilesTableRow.tsx:33]: How the invalid date looks like? Because if the data does not exists it is already displaying `--`

File: frontend/src/pages/acceleratorProfiles/components/TableData.tsx (hunk 5 above)
Comment (yannnz) [reply frontend/src/pages/acceleratorProfiles/components/TableData.tsx:9]: One thing I am not sure for the first option is, will it only display one line?  I guess what Gage saying is to display certain amount of characters of the description, so it might has two or three lines, rather than just limit the length in one line?

Comment (dpanshug) [reply frontend/src/pages/acceleratorProfiles/components/TableData.tsx:9]: Yes, the first option will only display it in one line, it will truncate content based on width and display the whole content using tooltip.

Comment (yannnz) [reply frontend/src/pages/acceleratorProfiles/components/TableData.tsx:9]: Oh I see. I didn't notice that. Is there a way to wrap the text and also have the truncate?  For instance, we can limit the characters of name to 64, wrap it. Limit the description to certain amount, only display 64 and wrap it, and truncate the rest. I need to check the convention of other tables.

Comment (dpanshug) [reply frontend/src/pages/acceleratorProfiles/components/TableData.tsx:9]: There was a [discussion](https://github.com/opendatahub-io/odh-dashboard/pull/1718#issuecomment-1708332464) before on truncating text under table, and it was decided to truncate text instead of wrapping it to the next line. 

cc @vconzola

Comment (Gkrumbach07) [reply frontend/src/pages/acceleratorProfiles/components/TableData.tsx:9]: Based on vinces comment, we should just limit width and truncate. but make the case where both the description and the name are long, both have their own tooltips.

However this is a different cases, because this is the description. title should not wrap, but description could be. argued to wrap (with some truncating too) in my opinion

Comment (yannnz) [reply frontend/src/pages/acceleratorProfiles/components/TableData.tsx:9]: @dpanshug I checked with the OpenShift AI UXD team, the feedback is to have the name limited into one line and the description limited into two lines, and truncate the rest. I think that is also the strategy in other places.

File: frontend/src/pages/acceleratorProfiles/screens/list/AcceleratorProfilesTableRow.tsx (hunk 3 above)
Comment (Gkrumbach07) [reply frontend/src/pages/acceleratorProfiles/screens/list/AcceleratorProfilesTableRow.tsx:33]: It renders at the date.now()

Comment (Gkrumbach07) [reply frontend/src/pages/acceleratorProfiles/screens/list/AcceleratorProfilesTableRow.tsx:33]: So the issue is with the Timestamp component. see this description of default date: https://v4-archive.patternfly.org/v4/components/timestamp#default

TLDR it default to now if its not defined (or invalid)

to test if a date string is invalid you can use `Date. parse()` which will return `NaN` if it is invalid. Then you can use the `isNaN()` function to test that

Comment (Gkrumbach07) [reply frontend/src/pages/acceleratorProfiles/screens/list/AcceleratorProfilesTableRow.tsx:33]: Yeah looks like the mocks use a relative time anyway. so just use that function instead of timestmap.

File: frontend/src/pages/acceleratorProfiles/screens/list/AcceleratorProfileEnableToggle.tsx [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+        aria-label={label}
+        id={`${name}-enable-switch`}
+        isChecked={enabled}
+        isDisabled={inProgress}
+        onClick={() => {
+          if (enabled) {
+            setIsModalOpen(true);
+          } else {
+            setInProgress(true);
+            updateAcceleratorProfile(name, true).then(() => setInProgress(false));
```
Comment (Gkrumbach07) [inline frontend/src/pages/acceleratorProfiles/screens/list/AcceleratorProfileEnableToggle.tsx:31]: this should be {enabled: true} not just true

File: frontend/src/pages/acceleratorProfiles/screens/list/AcceleratorProfileEnableToggle.tsx [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+          }
+        }}
+      />
+      <DisableAcceleratorProfileModal
+        isOpen={isModalOpen}
+        onClose={(confirmStatus) => {
+          if (confirmStatus) {
+            updateAcceleratorProfile(name, false);
```
Comment (Gkrumbach07) [inline frontend/src/pages/acceleratorProfiles/screens/list/AcceleratorProfileEnableToggle.tsx:39]: same. here

File: frontend/src/pages/acceleratorProfiles/screens/list/AcceleratorProfilesTableRow.tsx [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+            <Truncate content={accelerator.spec.description} />
+          </Text>
+        )}
+      </TextContent>
+    </Td>
+    <Td dataLabel="Identifier">{accelerator.spec.identifier}</Td>
+    <Td dataLabel="Enable">
+      <AcceleratorProfileEnableToggle
+        enabled={accelerator.spec.enabled}
+        name={accelerator.spec.displayName}
```
Comment (Gkrumbach07) [inline frontend/src/pages/acceleratorProfiles/screens/list/AcceleratorProfilesTableRow.tsx:27]: needs to be metadata.name not displayName

File: frontend/src/pages/acceleratorProfiles/screens/list/AcceleratorProfileEnableToggle.tsx [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+import { Switch } from '@patternfly/react-core';
+import DisableAcceleratorProfileModal from '~/pages/acceleratorProfiles/screens/list/DisableAcceleratorProfileModal';
+import { updateAcceleratorProfile } from '~/services/acceleratorProfileService';
+
+type AcceleratorProfileEnableToggleProps = {
+  enabled: boolean;
+  name: string;
+};
+
+const AcceleratorProfileEnableToggle: React.FC<AcceleratorProfileEnableToggleProps> = ({
```
Comment (Gkrumbach07) [inline frontend/src/pages/acceleratorProfiles/screens/list/AcceleratorProfileEnableToggle.tsx:11]: this is not updating. follow how `BYONImageStatusToggle` does this

File: frontend/src/pages/acceleratorProfiles/screens/list/AcceleratorProfilesTableRow.tsx [hunk 12]
```diff
@@ -0,0 +1,56 @@
...
+      ) : (
+        '--'
+      )}
+    </Td>
+    <Td isActionCell>
+      <ActionsColumn
+        items={[
+          {
+            title: 'Edit',
+            onClick: () => undefined,
```
Comment (Gkrumbach07) [inline frontend/src/pages/acceleratorProfiles/screens/list/AcceleratorProfilesTableRow.tsx:44]: edit should navigate to `/acceleratorProfiles/edit/{name}`

File: frontend/src/pages/acceleratorProfiles/screens/list/AcceleratorProfilesTableRow.tsx [hunk 13]
```diff
@@ -0,0 +1,56 @@
...
+  <Tr>
+    <Td dataLabel="Name">
+      <TextContent>
+        <Text>{accelerator.spec.displayName}</Text>
+        {accelerator.spec.description && (
+          <Text component={TextVariants.small}>
+            <Truncate content={accelerator.spec.description} />
+          </Text>
+        )}
+      </TextContent>
```
Comment (Gkrumbach07) [inline frontend/src/pages/acceleratorProfiles/screens/list/AcceleratorProfilesTableRow.tsx:21]: this needs to follow what yan said

> I checked with the OpenShift AI UXD team, the feedback is to have the name limited into one line and the description limited into two lines, and truncate the rest. The reason of the suggestion is that wrapping the description would cause different heights of the columns.

File: frontend/src/utilities/useAcceleratorState.ts [hunk 14]
```diff
@@ -100,6 +100,7 @@ const useAcceleratorState = (
                 kind: 'AcceleratorProfile',
                 metadata: {
                   name: 'migrated-gpu',
+                  namespace: 'opendatahub',
```
Comment (Gkrumbach07) [inline frontend/src/utilities/useAcceleratorState.ts:103]: this is not needed

**Review by Gkrumbach07 — CHANGES_REQUESTED** [review]

**Comment by Gkrumbach07** [comment]
@dpanshug I think the test is failing. it could be an aria issue

File: frontend/src/pages/acceleratorProfiles/screens/list/AcceleratorProfilesTableRow.tsx [hunk 15]
```diff
@@ -0,0 +1,77 @@
...
+
+const AcceleratorProfilesTableRow: React.FC<AcceleratorProfilesTableRow> = ({ accelerator }) => {
+  const navigate = useNavigate();
+  const modifiedDate = accelerator.metadata.annotations?.['opendatahub.io/modified-date'];
+
+  return (
+    <Tr>
+      <Td dataLabel="Name">
+        <TextContent>
+          <Text>{accelerator.spec.displayName}</Text>
```
Comment (Gkrumbach07) [inline frontend/src/pages/acceleratorProfiles/screens/list/AcceleratorProfilesTableRow.tsx:28]: make sure this also truncates

File: frontend/src/pages/acceleratorProfiles/screens/list/AcceleratorProfilesTableRow.tsx [hunk 16]
```diff
@@ -0,0 +1,77 @@
...
+                variant: TimestampTooltipVariant.default,
+              }}
+            >
+              {relativeTime(Date.now(), new Date(modifiedDate).getTime())}
+            </Timestamp>
+          </span>
+        ) : (
+          '--'
+        )}
+      </Td>
```
Comment (Gkrumbach07) [inline frontend/src/pages/acceleratorProfiles/screens/list/AcceleratorProfilesTableRow.tsx:58]: when modified date is invalid, it says just now, when really it should be `--`

also remove the style and span. a note on `style=`. this is essentially custom css and we should avoid it if we can

```suggestion
      <Td dataLabel="Last modified">
        {modifiedDate && !isNaN(new Date(modifiedDate).getTime()) ? (
          <Timestamp
            date={new Date(modifiedDate)}
            tooltip={{
              variant: TimestampTooltipVariant.default,
            }}
          >
            {relativeTime(Date.now(), new Date(modifiedDate).getTime())}
          </Timestamp>
        ) : (
          '--'
        )}
      </Td>
```

**Review by Gkrumbach07 — COMMENTED** [review]
added a few comments

**Review by Gkrumbach07 — APPROVED** [review]
/lgtm

**Comment by Gkrumbach07** [comment]
checked over once more. all looks good
